File: app/src/main/java/ru/warfare/darkannihilation/enemy/boss/DeathStar.java
package ru.warfare.darkannihilation.enemy.boss;

import ru.warfare.darkannihilation.arts.ImageHub;
import ru.warfare.darkannihilation.audio.AudioHub;
import ru.warfare.darkannihilation.base.BaseBoss;
import ru.warfare.darkannihilation.base.Sprite;
import ru.warfare.darkannihilation.bullet.BulletBoss;
import ru.warfare.darkannihilation.enemy.TripleFighter;
import ru.warfare.darkannihilation.math.Randomize;
import ru.warfare.darkannihilation.systemd.Game;

import static ru.warfare.darkannihilation.constant.Constants.BOSS_HEALTH;
import static ru.warfare.darkannihilation.constant.Constants.BOSS_SHOOT_TIME;
import static ru.warfare.darkannihilation.constant.Constants.NUMBER_VADER;
import static ru.warfare.darkannihilation.systemd.service.Windows.SCREEN_WIDTH;

public class DeathStar extends BaseBoss {
    public DeathStar(Game g) {
        super(g, ImageHub.bossImage, BOSS_HEALTH, 10, BOSS_SHOOT_TIME);

        recreateRect(x + 20, y + 20, right() - 20, bottom() - 20);
    }

    @Override
    public void shoot() {
        int ri = right() - 115;
        int y40 = y + 40;
        for (int i = 1; i < 4; i++) {
            game.intersectOnlyPlayer.add(new BulletBoss(game, ri, y40, i));
        }
        AudioHub.playShoot();
    }

    @Override
    public void kill() {
        Game.score += 325;

        int len = NUMBER_VADER / 4;
        for (int i = 0; i < len; i++) {
            if (Randomize.randFloat() <= 0.1) {
                game.enemies.add(new TripleFighter(game));
            } else {
                game.generateVader();
            }
        }
    }

    @Override
    public Sprite getRect() {
        return newRect(x + 20, y + 20);
    }

    @Override
    public void update() {
        if (y >= 35) {
            x += speedX;
            if (x < -width) {
                if (Randomize.randBoolean()) {
                    speedX = -speedX;
                } else {
                    x = SCREEN_WIDTH;
                }
            }
            if (x > SCREEN_WIDTH) {
                if (Randomize.randBoolean()) {
                    speedX = -speedX;
                } else {
                    x = -width;
                }
            }
            if (!startShoot) {
                startShoot = true;
                gameTask.start();
            }

        } else {
            super.update();

            y += speedY;
        }
    }

    @Override
    public void render() {
        Game.canvas.drawBitmap(image, x, y, null);

        super.render();
    }
}

File: app/src/main/java/ru/warfare/darkannihilation/enemy/boss/DeathStar.kt
package ru.warfare.darkannihilation.enemy.boss

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import ru.warfare.darkannihilation.*
import ru.warfare.darkannihilation.Constants.BOSS_HEALTH
import ru.warfare.darkannihilation.Constants.BOSS_SHOOT_TIME
import ru.warfare.darkannihilation.base.Sprite
import ru.warfare.darkannihilation.bullet.BulletBoss
import ru.warfare.darkannihilation.hub.AudioHub
import ru.warfare.darkannihilation.hub.ImageHub
import ru.warfare.darkannihilation.math.Math
import ru.warfare.darkannihilation.systemd.Game

class DeathStar(game: Game) : Sprite(game, ImageHub.bossImage) {
    private var lastShoot = System.currentTimeMillis()
    private var now: Long = 0
    private var hp: Float = 10f

    init {
        health = BOSS_HEALTH.toInt()
        speedY = 1
        speedX = 10
        isPassive = true
        x = Math.randInt(0, screenWidthWidth)
        y = -800
        ImageHub.loadPortalImages()
        recreateRect(x + 20, y + 20, right() - 20, bottom() - 20)
    }

    private fun shoot() {
        now = System.currentTimeMillis()
        if (now - lastShoot > BOSS_SHOOT_TIME) {
            runBlocking {
                launch(Dispatchers.IO) {
                    lastShoot = now
                    val r = right() - 115
                    val y20 = y + 40
                    for (i in 1..3) {
                        Game.allSprites.add(
                            BulletBoss(
                                r,
                                y20,
                                i
                            )
                        )
                    }
                    AudioHub.playShoot()
                }
            }
        }
    }

    private fun killAfterFight() {
        if (HardThread.job == 0) {
            HardThread.job = 10
            createSkullExplosion()
            Game.bosses.remove(this)
            Game.allSprites.remove(this)
        }
    }

    override fun getRect(): Sprite {
        return newRect(x + 20, y + 20)
    }

    override fun check_intersectionBullet(bullet: Sprite) {
        if (intersect(bullet)) {
            health -= bullet.damage
            bullet.intersection()
        }
    }

    override fun update() {
        if (y >= 35) {
            x += speedX
            if (x < -width) {
                if (Math.randInt(1, 2) == 1) {
                    speedX = -speedX
                } else {
                    x = Game.screenWidth
                }
            }
            if (x > Game.screenWidth) {
                if (Math.randInt(1, 2) == 1) {
                    speedX = -speedX
                } else {
                    x = -width
                }
            }
            shoot()

            if (health <= 0) {
                killAfterFight()
            }
        } else {
            if (y == -600) {
                AudioHub.restartBossMusic()
                AudioHub.pauseBackgroundMusic()
                Game.gameStatus = 5
            }
            y += speedY
        }
    }

    override fun render() {
        Game.canvas.drawBitmap(image, x.toFloat(), y.toFloat(), null)

        hp = if (hp > 4) {
            ((health / BOSS_HEALTH.toFloat()) * 140)
        } else {
            4f
        }
        Game.canvas.drawRect(
            (centerX() - 70).toFloat(),
            (y - 10).toFloat(),
            (centerX() + 70).toFloat(),
            (y + 5).toFloat(),
            Game.scorePaint
        )
        Game.canvas.drawRect(
            (centerX() - 68).toFloat(),
            (y - 8).toFloat(),
            (centerX() - 72 + hp),
            (y + 3).toFloat(),
            Game.topPaintRed
        )
    }
}